File: worker/src/stats.ts
export interface Turn {
  speaker: "interviewer" | "candidate"
  text: string
  startMs: number
  endMs: number
}

const FILLERS = /\b(um+|uh+|erm|you know|basically|kind of|sort of|i mean)\b|\blike,/gi

export function candidateStats(turns: Turn[]) {
  const mine = turns.filter((t) => t.speaker === "candidate")
  const words = mine.reduce((n, t) => n + (t.text.trim().match(/\S+/g)?.length ?? 0), 0)
  const speakingMs = mine.reduce((n, t) => n + Math.max(0, t.endMs - t.startMs), 0)
  const fillers = mine.reduce((n, t) => n + (t.text.match(FILLERS)?.length ?? 0), 0)
  const wpm = speakingMs > 5_000 ? Math.round(words / (speakingMs / 60_000)) : 0
  return { words, wpm, fillers }
}

export function clock(seconds: number) {
  const s = Math.max(0, Math.round(seconds))
  return `${String(Math.floor(s / 60)).padStart(2, "0")}:${String(s % 60).padStart(2, "0")}`
}

File: android/app/src/main/java/dev/mintu/hotseat/live/Stats.kt
package dev.mintu.hotseat.live

// same counting as worker/src/stats.ts so the live caption matches the report
private val FILLERS = Regex("""\b(um+|uh+|erm|you know|basically|kind of|sort of|i mean)\b|\blike,""", RegexOption.IGNORE_CASE)

data class Pace(val words: Int, val wpm: Int, val fillers: Int)

fun candidatePace(turns: List<Turn>): Pace {
    val mine = turns.filter { it.speaker == Speaker.candidate }
    val words = mine.sumOf { t -> t.text.trim().split(Regex("""\s+""")).count { it.isNotEmpty() } }
    val speakingMs = mine.sumOf { maxOf(0L, it.endMs - it.startMs) }
    val fillers = mine.sumOf { FILLERS.findAll(it.text).count() }
    val wpm = if (speakingMs > 5_000) Math.round(words / (speakingMs / 60_000.0)).toInt() else 0
    return Pace(words, wpm, fillers)
}

fun clock(ms: Long): String {
    val s = maxOf(0L, ms) / 1000
    return "%02d:%02d".format(s / 60, s % 60)
}
